### ISSUE TITLE
Do you know if emayili works with Microsoft365R?

### ISSUE THREAD

**Issue by smach** [datawookie/emayili#99]
Do you know if rendering R Markdown works with the [Microsoft365R package](https://github.com/Azure/Microsoft365R)? I've also asked on their GitHub, since they say their package works with emayili, but I can't get it to work even with basic html.  I figured it was also worth asking you in case you knew. Code in their unit tests of

```
    ey_em <- emayili::envelope(
        to=to_addr,
        subject="test emayili email",
        html="test emayili email"
    )
    em <- folder$create_email(ey_em)
```

gives me this error:

`Error: $ operator is invalid for atomic vectors`

when I try even with an HTML file and not R Markdown. It's possible my problem is that I'm using their authentication for business Outlook instead of personal Outlook, and maybe it works with personal. Anyway, figured I'd ask if you ever tested it. And thanks very much for the cool package!

**Comment by datawookie** [comment]
Hello Sharon, I'm Andrew or datawookie.

**Comment by datawookie** [comment]
Hi @smach,

> gives me this error

Your code has two distinct statements. Can you please tell me which of them is generating the error?

I assume that it's not this:

```
    ey_em <- emayili::envelope(
        to=to_addr,
        subject="test emayili email",
        html="test emayili email"
    )
```

because that's a perfectly standard call to `{emayili}`.

What is `folder`? Without that context it's difficult to know where the problem might be.

Thanks, Andrew.

**Comment by smach** [comment]
Hi Andrew. I watched your BirminghamR video last night and was hoping I might be able to figure out how to use this with corporate Outlook. Sorry, "folder" is the Microsoft365R object that authenticates with your Outlook, and then you should be able to send the email message with the send method. Probably a question for them. I thought you might know about it because Microsoft Azure says their package works with emayili ("Microsoft365R also supports using the blastula and emayili packages to write nicely-formatted HTML emails. In particular, blastula lets you compose your email in Markdown, which is often easier than writing raw HTML.")

**Comment by datawookie** [comment]
Hi Sharon,

Aha! Okay, I found the bit of code in the tests that you were working from.

Here's a slightly more complete example for reference:

```r
library(Microsoft365R)
library(emayili)

outlook <- get_personal_outlook()

folder <- outlook$create_folder("test-emayili")

body <- envelope(
  to = "bob@gmail.com",
  subject = "test emayili email"
) %>%
  html("<p>Hello!</p>")

email <- folder$create_email(body)

email$send()

folder$delete(confirm = FALSE)
```

However, unfortunately, the support in `{Microsoft365R}` is based on a very old version of `{emayili}`. A bit of work would be required to get it compatible again. 

I feel that the API of `{emayili}` has now stabilised, so it would be a reasonable thing to do at this stage.

Thanks, Andrew.

**Comment by jonspring** [comment]
Hi @datawookie and @hongooi73, I followed the thread above and https://github.com/Azure/Microsoft365R/issues/95 with some eagerness, as I would like to render Rmd files with emayil, and send that as inline HTML through a business Outlook account via Microsoft365R. I'm not sure which of the packages (or both, or neither!) is the source of my issue. 

When I ran `test-compatibility-microsoft365r.R`, I got the expected results, with the formatted body text and attachment.

If I knit RStudio's Rmd template to html ("template.html"), I can attach it to an email and send, but it arrives as an unlabeled "Attachment" with File Type "File":

```
outlb <- get_business_outlook()
my_envelope <- envelope(
  from = "foo@baz.org", to = "foo@baz.org",
  subject = "Thank you datawookie and hongooi73") %>%
  attachment("template.html",  type = "text/html")
msg <- outlb$create_email(my_envelope)
msg$send()
```
If I add `disposition = "inline"` to the `attachment()` call, the file is now called "inline" and the body of the email seems to show the contents as Base64 encoding.

If I replace the `attachment()` line above with `render("template.Rmd", include_css = FALSE)`, the html components of `my_envelope` appear to be successfully generated (see below), but when I run `msg <- outlb$create_email(my_envelope); msg$send()`, the email has an empty body and is only 10 KB. 

Using R 4.2.2 on Windows 10, emayili 0.7.13, and Microsoft365R 2.3.4.
Using Microsoft Office 365 ProPlus, Outlook version 2002.


```
Date:                         Sat, 25 Feb 2023 07:44:58 GMT
X-Mailer:                     {emayili}-0.7.13
MIME-Version:                 1.0
To:                           foo@baz.org
From:                         foo@baz.org
Subject:                      Thank you datawookie and hongooi73
Content-Type:                 multipart/related; 
                              boundary="0b65f749"

--0b65f749
Content-Type:                 text/html; 
                              charset=utf-8

<html>
<head>
<title>example</title>
</head>
<body>
<div class="container-fluid main-container">
<div id="header">
<h1 class="title toc-ignore">example</h1>
<h4 class="date">2023-02-23</h4>
</div>
<div id="r-markdown" class="section level2">
<h2>R Markdown</h2>
<p>This is an R Markdown document. Markdown is a simple formatting
syntax for authoring HTML, PDF, and MS Word documents. For more details
on using R Markdown see <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a>.</p>
<p>When you click the <strong>Knit</strong> button a document will be
generated that includes both content as well as the output of any
embedded R code chunks within the document. You can embed an R code
chunk like this:</p>
<pre class="r"><code>summary(cars)</code></pre>
<pre><code>##      speed           dist       
##  Min.   : 4.0   Min.   :  2.00  
##  1st Qu.:12.0   1st Qu.: 26.00  
##  Median :15.0   Median : 36.00  
##  Mean   :15.4   Mean   : 42.98  
##  3rd Qu.:19.0   3rd Qu.: 56.00  
##  Max.   :25.0   Max.   :120.00</code></pre>
</div>
<div id="including-plots" class="section level2">
<h2>Including Plots</h2>
<p>You can also embed plots, for example:</p>
<figure><img src="cid:73317b82" width="672"></figure><p>Note that the <code>echo = FALSE</code> parameter was added to the
code chunk to prevent printing of the R code that generated the
plot.</p>
</div>
</div>
</body>
</html>
--0b65f749
Content-Type:                 image/png; 
                              name="file15b41db5b37.png"
Content-Disposition:          inline; 
                              filename="file15b41db5b37.png"
Content-Transfer-Encoding:    base64
X-Attachment-Id:              73317b82
Content-ID:                   <73317b82>
Content-MD5:                  UANFmxRdTS0dS7AScA+YbQ==

iVBORw0KGgoAAAANSUhEUgAABUAAAAPACAMAAADDuCPrAAAA1VBMVEUAAAAAADoAAGYAOjoAOmYA
OpAAZpAAZrY6AAA6OgA6Ojo6OmY6ZmY6ZpA6ZrY6kLY6kNtmAABmADpmOgBmOjpmZjpmZmZmZpBm
kLZmkNtmtrZmtttmtv+QOgCQOjqQZgCQZjqQZmaQkJCQkLaQtraQttuQ29uQ2/+2ZgC2Zjq2kDq2
kGa2kJC2tpC2tra2ttu227a229u22/+2///bkDrbkGbbtmbbtpDbtrbb25Db27bb29vb2//b/7bb
/9vb////tmb/25D/27b/29v//7b//9v///9QcbwyAAAACXBIWXMAAB2HAAAdhwGP5fFlAAAgAElE
QVR4nO3dbWPT1rqgYYWGwYVNDwxtyh5gDj0HpmEzTYc2hR7oDCHF+f8/aSS/RX5JYj1Z0dJyrusL
wSB5ocQ3kiUtV2cAhFS5BwBQKgEFCBJQgCABBQgSUIAgAQUIElCAIAEFCBJQgCABBQgSUIAgAQUI
...
...
```

**Comment by datawookie** [comment]
> If I knit RStudio's Rmd template to html ("template.html"), I can attach it to an email and send, but it arrives as an unlabeled attachment.

This is the expected behaviour. It's an attachment just like any other file.

> If I add disposition = "inline" to the attachment() call, the file is now called "inline" and the body of the email seems to show the contents as Base64 encoding.

This is the behaviour with an _inline attachment_ (it's still an attachment).

> the html components appear to be successfully generated

Indeed, the HTML components _are_ successfully generated.

Let's get rid of the complexity and look at a simple example.

```r
msg <- envelope(
  from = "anne@example.com",
  to = "bob@example.com",
  subject = "Render"
) %>%
  render("This is the body.")
```

This is what the resulting message looks like:

```r
print(msg, details=T)
```
```
Date:                         Sun, 26 Feb 2023 07:17:11 GMT
X-Mailer:                     {emayili}-0.7.14
MIME-Version:                 1.0
To:                           bob@example.com
From:                         anne@example.com
Subject:                      Render
Content-Type:                 text/html; 
                              charset=utf-8

<html><body><p>This is the body.</p></body></html>
```

This is a valid email message with HTML in the body.

Now convert to something that `{Microsoft365R}` understands.

```r
msg_outlook <- outlook$create_email(msg)
```

Take a look at that.

```
<Outlook email>
  directory id: AAkALgAAAAAAHYQDEapmEc2byACqAC-EWg0AVS2PvDzLYkCjNblTRE4JSgAFbJpc6wAA 
  from:
  sent:
  to: bob@example.com 
  subject: Render 
---
This is the body.
```

Again, this looks legit, but  @hongooi73 would need to confirm. I see that the `From` field isn't populated, but I presume that this is filled in  when the message is actually dispatch.

TL/DR: I don't think that the problem lies with `{emayili}`. It is generating a valid email with rendered HTML content. I believe that the problem arises when translating this across to `{Microsoft365R}`. I'm very happy to work with @hongooi73 to smooth this process. I don't have a Microsoft365 account, so this is hard for me to test.

**Comment by jonspring** [comment]
Thank you for taking a look. Happy to try any tests if that would be helpful.
